kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/TokenExpandingJsonFactoryWrapperTest.java

```diff
@@ -0,0 +1,155 @@
+    };
+    private static final TypeReference<String> SCALAR_TYPE_REF = new TypeReference<>() {
+    };
+
+    static Stream<JsonFactory> factories() {
+        return Stream.of(new YAMLFactory(), new JsonFactory());
+    }
+
+    @ParameterizedTest
+    @MethodSource(value = "factories")
+    void wrappedFactoryIsSameType(JsonFactory f) {
+        var wrapped = wrap(f);
+        assertThat(wrapped)
+                .isNotNull()
+                .isInstanceOf(f.getClass());
+    }
+
+    static Stream<Arguments> deserializationInput() {
+        return Stream.of(
+                Arguments.of("yaml - obj - 1 field - no expansion", new YAMLFactory(), """
+                        myString: foo
+                        """, OBJECT_TYPE_REF, new SimpleBean("foo", false, 0)),
+                Arguments.of("yaml - obj - 3 fields - no expansion", new YAMLFactory(), """
+                        myString: foo
+                        myBoolean: true
+                        myInt: 1
+                        """, OBJECT_TYPE_REF, new SimpleBean("foo", true, 1)),
+                Arguments.of("yaml - obj - 1 field - expanded", new YAMLFactory(), """
+                        myString: ${sys:foostr}
+                        """, OBJECT_TYPE_REF, new SimpleBean("bar", false, 0)),
+                Arguments.of("yaml - obj - 3 fields - expanded", new YAMLFactory(), """
+                        myString: ${sys:foostr}
+                        myBoolean: ${sys:foobool}
+                        myInt: ${sys:fooint}
+                        """, OBJECT_TYPE_REF, new SimpleBean("bar", true, 5)),
+                Arguments.of("yaml - array of objects", new YAMLFactory(), """
+                        - myString: ${sys:foostr}
+                          myBoolean: ${sys:foobool}
+                          myInt: ${sys:fooint}
+                        - myString: foo
+                          myBoolean: false
+                          myInt: 4
+                        """, ARRAY_OF_OBJECTS_TYPE_REF, new SimpleBean[]{ new SimpleBean("bar", true, 5), new SimpleBean("foo", false, 4) }),
+                Arguments.of("yaml - array of scalars", new YAMLFactory(), """
+                        - ${sys:foostr}
+                        - foo
+                        """, ARRAY_OF_SCALARS_TYPE_REF, new String[]{ "bar", "foo" }),
+                Arguments.of("yaml - scalar - expanded", new YAMLFactory(), """
+                        ${sys:foostr}
+                        """, SCALAR_TYPE_REF, "bar"),
+                Arguments.of("json - obj - 3 fields - expanded", new JsonFactory(), """
+                        {
+                            "myString": "${sys:foostr}",
+                            "myBoolean": "${sys:foobool}",
+                            "myInt": "${sys:fooint}"
+                        }
+                        """, OBJECT_TYPE_REF, new SimpleBean("bar", true, 5))
+
+        );
```

Comment:
Maybe I'm missing it, but what happens if you have a bean property of type `int` or `boolean` but the looked up value is incompatible? I'm sure it will throw an exception, but how understandable is the exception message in this case, and can it leak sensitive data?

Similarly, if I have a Java bean property of type string and I pass something that's syntactically a boolean or an int. 

We should also have cases covering `null`: Can an Interpolator interpolate a `null` value?